File: core/src/main/java/xtdb/trie/ArrowHashTrie.java
package xtdb.trie;

import org.apache.arrow.vector.IntVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.complex.DenseUnionVector;
import org.apache.arrow.vector.complex.ListVector;
import org.apache.arrow.vector.complex.StructVector;

import java.util.stream.IntStream;

public class ArrowHashTrie implements HashTrie<ArrowHashTrie.Node> {

    private static final byte BRANCH_TYPE_ID = 1;
    private static final byte LEAF_TYPE_ID = 2;

    private final DenseUnionVector nodesVec;
    private final ListVector branchVec;
    private final IntVector branchElVec;
    private final StructVector leafVec;
    private final IntVector pageIdxVec;

    private ArrowHashTrie(VectorSchemaRoot trieRoot) {
        nodesVec = ((DenseUnionVector) trieRoot.getVector("nodes"));
        branchVec = (ListVector) nodesVec.getVectorByType(BRANCH_TYPE_ID);
        branchElVec = (IntVector) branchVec.getDataVector();
        leafVec = (StructVector) nodesVec.getVectorByType(LEAF_TYPE_ID);
        pageIdxVec = leafVec.getChild("page-idx", IntVector.class);
    }

    public sealed interface Node extends HashTrie.Node<Node> {
    }

    private static byte[] conjPath(byte[] path, byte idx) {
        int currentPathLength = path.length;
        var childPath = new byte[currentPathLength + 1];
        System.arraycopy(path, 0, childPath, 0, currentPathLength);
        childPath[currentPathLength] = idx;
        return childPath;
    }

    public final class Branch implements Node {

        private final byte[] path;
        private final int branchVecIdx;

        public Branch(byte[] path, int branchVecIdx) {
            this.path = path;
            this.branchVecIdx = branchVecIdx;
        }

        public Node[] children() {
            int startIdx = branchVec.getElementStartIndex(branchVecIdx);

            return IntStream.range(0, branchVec.getElementEndIndex(branchVecIdx) - startIdx)
                    .mapToObj(childBucket -> {
                        int childIdx = childBucket + startIdx;
                        return branchElVec.isNull(childIdx) ? null : forIndex(conjPath(path, (byte) childBucket), branchElVec.get(childIdx));
                    })
                    .toArray(Node[]::new);
        }

        @Override
        public byte[] path() {
            return path;
        }
    }

    public final class Leaf implements Node {

        private final byte[] path;
        private final int leafVecIdx;

        public Leaf(byte[] path, int leafVecIdx) {
            this.path = path;
            this.leafVecIdx = leafVecIdx;
        }

        public int getPageIndex() {
            return pageIdxVec.get(leafVecIdx);
        }

        @Override
        public byte[] path() {
            return path;
        }

        @Override
        public Node[] children() {
            return null;
        }
    }

    private Node forIndex(byte[] path, int idx) {
        var nodeOffset = nodesVec.getOffset(idx);

        return switch (nodesVec.getTypeId(idx)) {
            case 0 -> null;
            case BRANCH_TYPE_ID -> new Branch(path, nodeOffset);
            case LEAF_TYPE_ID -> new Leaf(path, nodeOffset);
            default -> throw new UnsupportedOperationException();
        };
    }

    public static HashTrie<Node> from(VectorSchemaRoot root) {
        return new ArrowHashTrie(root);
    }

    @Override
    public HashTrie.Node<Node> rootNode() {
        return forIndex(new byte[0], nodesVec.getValueCount() - 1);
    }
}

File: core/src/main/kotlin/xtdb/trie/ArrowHashTrie.kt
package xtdb.trie

import xtdb.arrow.Vector

private const val NIL = "nil"
private const val BRANCH_IID = "branch-iid"
private const val LEAF = "leaf"
private const val DATA_PAGE_IDX = "data-page-idx"

class ArrowHashTrie(private val nodesVec: Vector) : HashTrie<ArrowHashTrie.Node, ArrowHashTrie.Leaf> {

    private val iidBranchVec = nodesVec[BRANCH_IID]
    private val iidBranchElVec = iidBranchVec.listElements

    private val dataPageIdxVec = nodesVec[LEAF][DATA_PAGE_IDX]

    interface Node : HashTrie.Node<Node>

    inner class IidBranch(override val path: ByteArray, branchVecIdx: Int) : Node {
        private val startIdx = iidBranchVec.getListStartIndex(branchVecIdx)
        private val count = iidBranchVec.getListCount(branchVecIdx)

        override val hashChildren: List<Node?>
            get() = List(count) { childBucket ->
                val childIdx = childBucket + startIdx
                if (iidBranchElVec.isNull(childIdx))
                    null
                else
                    forIndex(conjPath(path, childBucket.toByte()), iidBranchElVec.getInt(childIdx))
            }
    }

    class Leaf(override val path: ByteArray, val dataPageIndex: Int) : Node {
        override val hashChildren = null
    }

    private fun forIndex(path: ByteArray, idx: Int) =
        when (nodesVec.getLeg(idx)) {
            NIL -> null
            BRANCH_IID -> IidBranch(path, idx)
            LEAF -> Leaf(path, dataPageIdxVec.getInt(idx))
            else -> error("unknown leg: ${nodesVec.getLeg(idx)}")
        }

    override val rootNode get() = forIndex(ByteArray(0), nodesVec.valueCount - 1)
}

